homeassistant/components/panasonic_viera/__init__.py

```diff
@@ -115,71 +137,99 @@ class Remote:
+
+        self._state = STATE_OFF
+        self._app_info = None
+        self._volume = "0"
+        self._mute = "0"
+
+        self.playing = False
 
         self._control = None
+        self._timeout = 5
 
     async def async_create_remote_control(self, during_setup=False):
         """Create remote control."""
-        control_existed = self._control is not None
         try:
             params = {}
             if self._app_id and self._encryption_key:
                 params["app_id"] = self._app_id
                 params["encryption_key"] = self._encryption_key
 
-            self._control = await self._hass.async_add_executor_job(
-                partial(RemoteControl, self._host, self._port, **params)
+            control = await self._hass.async_add_executor_job(
+                partial(
+                    RemoteControl,
+                    self._host,
+                    self._port,
+                    listen_port=self._listen_port,
+                    **params,
+                )
             )
 
-            self.state = STATE_ON
+            await control.async_start_server()
+            control.on_event = self.on_event
+
+            for event in UPNP_EVENTS:
+                await self._hass.async_add_executor_job(
+                    partial(control.upnp_service_subscribe, event)
+                )
+
+            self._control = control
+
             self.available = True
+            self.connected = True
         except (TimeoutError, URLError, SOAPError, OSError) as err:
-            if control_existed or during_setup:
-                _LOGGER.debug("Could not establish remote connection: %s", err)
+            if during_setup:
+                _LOGGER.error("Could not establish remote connection: %s", err)
 
-            self._control = None
-            self.state = STATE_OFF
             self.available = self._on_action is not None
+            self.connected = False
         except Exception as err:  # pylint: disable=broad-except
-            if control_existed or during_setup:
-                _LOGGER.exception("An unknown error occurred: %s", err)
-                self._control = None
-                self.state = STATE_OFF
-                self.available = self._on_action is not None
+            _LOGGER.exception("An unknown error occurred: %s", err)
+            self.available = self._on_action is not None
+            self.connected = False
 
     async def async_update(self):
         """Update device data."""
-        if self._control is None:
+        if not self.connected:
             await self.async_create_remote_control()
             return
 
         await self._handle_errors(self._update)
 
     def _update(self):
-        """Retrieve the latest data."""
-        self.muted = self._control.get_mute()
-        self.volume = self._control.get_volume() / 100
+        for event in UPNP_EVENTS:
+            self._control.upnp_service_resubscribe(event, self._timeout)
```

Comment:
Is there a way of making this run outside of update, in a separate schedule? Like, every two minutes?